packages/reflow/src/ViewProxy.ts

```diff
@@ -1,82 +1,100 @@
-export type ViewOnRemoveCallback = () => void;
-
-export class ViewProxy<ViewsMap extends ViewsMapInterface, T extends ViewsMap[keyof ViewsMap]> extends Promise<T["output"]> {
-	private doneCalled = false;
-	private doneCallValue: T["output"];
-	private removed: boolean = false;
-	private eventListeners: {
-		[U in ViewInterfaceEvents<ViewsMap, T>]?: Array<ViewInterfaceEventCallback<ViewsMap, T, U>>
-	};
-	private resolve: (output: T["output"]) => void;
-	private reject: (output: Error) => void = (output) => { };
-	private onUpdate: ViewOnUpdateCallback<ViewsMap, T> = (input) => { };
-	private onRemove: ViewOnRemoveCallback = () => { };
-
-	constructor(executor: (resolve: () => void, reject: () => void) => ViewProxy<ViewsMap, T> | null, onUpdate: ViewOnUpdateCallback<ViewsMap, T>, onRemove: ViewOnRemoveCallback) {
-		super(executor ? executor : (resolve) => {
-			setTimeout(() => {
-				this.resolve = resolve;
-				// technically, because we're assigning this.resolve in a timeout, ViewProxy.done can be called synchronously before the timer is done.
-				// in such case, ViewProxy.done flags it was called, and keep the output value - we'll immediately resolve
-				if (this.doneCalled) {
-					resolve(this.doneCallValue);
-				}
-			});
-		});
-		this.onUpdate = onUpdate || this.onUpdate;
-		this.onRemove = onRemove || this.onRemove;
-		this.eventListeners = {};
-	}
-	done(output: T["output"]) {
-		if (!this.resolve) {
-			this.doneCalled = true;
-			this.doneCallValue = output;
-			return;
-		}
-		this.resolve(output);
-	}
-	event<U extends ViewInterfaceEvents<ViewsMap, T>>(eventName: U, data: ViewInterfaceEventData<ViewsMap, T, U>) {
-		if (this.removed) {
-			return;
-		}
-		if (!this.eventListeners[eventName]) {
-			return;
-		}
-		for (const listener of this.eventListeners[eventName]) {
-			listener(data);
-		}
-	}
-	on<U extends ViewInterfaceEvents<ViewsMap, T>>(eventName: U, listener: ViewInterfaceEventCallback<ViewsMap, T, U>): ViewProxy<ViewsMap, T> {
-		if (this.removed) {
-			return;
-		}
-		if (!this.eventListeners[eventName]) {
-			this.eventListeners[eventName] = [];
-		}
-		this.eventListeners[eventName].push(listener);
-		return this;
-	}
-	update(params: PartialViewInterfaceInput<ViewsMap, T>) {
-		if (this.removed) {
-			return;
-		}
-		this.onUpdate(params);
-	}
-	remove() {
-		this.removed = true;
-		this.onRemove();
-		// when a view is removed, resolve the promise.
-		this.done(undefined);
-	}
-}
+import { ViewInterface, ViewsMapInterface } from "./View";
+
+export type PartialViewInterfaceInput<ViewsMap extends ViewsMapInterface, T extends ViewsMap[keyof ViewsMap]> = {
+	[U in keyof T["input"]]?: T["input"][U]
+};
+
+export type ParamsUnpack<T> = T extends (params: infer U) => any ? U :
+    T;
+
+export type ReturnUnpack<T> = T extends (params: any) => infer U ? U :
+	void;

```

Comment:
If i understand correctly, you use the function signature to infer the params, so you can conclude the sent data and the resolved value. There's no real need for the function here, right?
If this is the case, instead of a function signature, use a custom type exported from this file (expose it to the user through the index of course), and infer using it.
Let's say we introduce a type called `BiDirectionalEvent<Data, ResolvedValue>`
So and event definition can be:
``` ts
interface Events {
    event1: string;
    event2: BiDirectionalEvent<string, number>
}
```

Then you can infer using it:
``` ts
export type ParamsUnpack<T> = T extends BiDirectionalEvent<infer U, any> ? U : T
export type ReturnUnpack<T> = T extends BiDirectionalEvent<any, infer U> ? U : T
```

Would it work? seems cleaner to me

Reply:
I agree that inferring from a function can be quite confusing but it's still cool, maybe having support for both of them is the answer - something like
```typescript
type BiDirectionalEvent<Input, Output> = (input: Input) => Output;
``` 
what do you think? I personly think both ways have some pros and cons since using a function is cool but also a bit too tricky.